target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c

```diff
@@ -0,0 +1,244 @@
+ * Based on the sysfs hard config driver by Thibaut VARÈNE <hacks+kernel@slashdirt.org>
+ * Comments documenting the format carried over from routerboot.c
+ */
+
+#include <linux/etherdevice.h>
+#include <linux/mod_devicetable.h>
+#include <linux/mtd/mtd.h>
+#include <linux/nvmem-consumer.h>
+#include <linux/nvmem-provider.h>
+#include <linux/platform_device.h>
+
+#include "rb_hardconfig.h"
+
+#ifdef CONFIG_CPU_BIG_ENDIAN
+#define RB_MAGIC_HARD		"draH"
+#else
+#define RB_MAGIC_HARD		"Hard"
+#endif
+
+struct hard_config_header {
+	u8 magic[4];
+} __packed;
+
+/*
+ * Routerboot tag nodes are u32 values:
+ * - The low nibble is the tag identification number,
+ * - The high nibble is the tag payload length (node excluded) in bytes.
+ * Tag nodes are 32bit-aligned and in cpu-endian.
+ */
+#ifdef CONFIG_CPU_BIG_ENDIAN
+struct hard_config_tlv {
+	u16 len;
+	u16 tag;
+} __packed;
+#else
+struct hard_config_tlv {
+	u16 tag;
+	u16 len;
+} __packed;
+#endif
+
+static const char *rb_tlv_cell_name(u16 tag)
+{
+	switch (tag) {
+	case RB_ID_FLASH_INFO:
+		return "flash-info";
+	case RB_ID_MAC_ADDRESS_PACK:
+		return "base-mac-address";
+	case RB_ID_BOARD_PRODUCT_CODE:
+		return "board-product-code";
+	case RB_ID_BIOS_VERSION:
+		return "booter-version";
+	case RB_ID_SERIAL_NUMBER:
+		return "board-serial";
+	case RB_ID_MEMORY_SIZE:
+		return "mem-size";
+	case RB_ID_MAC_ADDRESS_COUNT:
+		return "mac-count";
+	case RB_ID_HW_OPTIONS:
+		return "hw-options";
+	case RB_ID_WLAN_DATA:
+		return "wlan-data";
+	case RB_ID_BOARD_IDENTIFIER:
+		return "board-identifier";
+	case RB_ID_PRODUCT_NAME:
+		return "product-name";
+	case RB_ID_DEFCONF:
+		return "defconf";
+	case RB_ID_BOARD_REVISION:
+		return "board-revision";
+	default:
+		break;
+	}
+
+	return NULL;
+}
+
+static int rb_tlv_mac_read_cb(void *priv, const char *id, int index,
+                              unsigned int offset, void *buf,
+                              size_t bytes)
+{
+	eth_addr_add(buf, index);
+
+	return 0;
+}
+
+static nvmem_cell_post_process_t rb_tlv_read_cb(u16 tag, u8 *buf)
+{
+	switch (tag) {
+	case RB_ID_MAC_ADDRESS_PACK:
+		return &rb_tlv_mac_read_cb;
+	default:
+		break;
+	}
+
+	return NULL;
+}
+
+static int rb_add_cells(struct device *dev, struct nvmem_device *nvmem,
+                        size_t data_len, u8 *data)
+{
+	struct nvmem_cell_info cell = {};
+	struct device_node *layout;
+	struct hard_config_tlv tlv;
+	u32 hdr_len = sizeof(struct hard_config_header);
+	u32 offset = 0;
+	int ret;
+
+	layout = of_nvmem_layout_get_container(nvmem);
+	if (!layout)
+		return -ENOENT;
+
+	/*
+	 * The payload immediately follows the tag node. Tag nodes are 32bit-aligned.
+ 	 * Payload offset will always be aligned. while length may not end on 32bit
+ 	 * boundary (the only known case is when parsing ERD data).
+ 	 * The nodes are cpu-endian on the flash media. The payload is cpu-endian when
+ 	 * applicable. Tag nodes are not ordered (by ID) on flash.
+ 	 */
+	while (offset < data_len) {
```

Comment:
```suggestion
	while ((offset + sizeof(tlv)) <= data_len) {
```
Should only be a problem with bad configuration or bad data, but want to make sure the type length read we do is still within data.